File: request-layer/server/src/limiter/rateLimiter.js
const config = require("../config");

/**
 * TokenBucketRateLimiter — Per-agent rate limiting using Redis-backed token buckets.
 *
 * Each agent gets its own bucket with configurable:
 *   - maxTokens (burst capacity)
 *   - refillRate (tokens per second)
 *
 * Uses a Lua script for atomic token operations (no race conditions).
 */
class RateLimiter {
  /**
   * @param {import("ioredis").Redis} redis
   */
  constructor(redis) {
    this.redis = redis;
    this.configKey = "ratelimit:config";

    // Lua script for atomic token bucket operation
    // Returns: [allowed (0|1), tokensRemaining, retryAfterMs]
    this.luaScript = `
      local key = KEYS[1]
      local maxTokens = tonumber(ARGV[1])
      local refillRate = tonumber(ARGV[2])
      local now = tonumber(ARGV[3])

      local bucket = redis.call('HMGET', key, 'tokens', 'lastRefill')
      local tokens = tonumber(bucket[1])
      local lastRefill = tonumber(bucket[2])

      -- Initialize bucket if new
      if tokens == nil then
        tokens = maxTokens
        lastRefill = now
      end

      -- Refill tokens based on elapsed time
      local elapsed = (now - lastRefill) / 1000
      local newTokens = elapsed * refillRate
      tokens = math.min(maxTokens, tokens + newTokens)

      -- Try to consume a token
      if tokens >= 1 then
        tokens = tokens - 1
        redis.call('HMSET', key, 'tokens', tokens, 'lastRefill', now)
        return {1, math.floor(tokens * 100) / 100, 0}
      else
        -- Calculate time until next token
        local deficit = 1 - tokens
        local retryAfterMs = math.ceil((deficit / refillRate) * 1000)
        redis.call('HMSET', key, 'tokens', tokens, 'lastRefill', now)
        return {0, 0, retryAfterMs}
      end
    `;
  }

  /**
   * Try to acquire a token for an agent.
   * @param {string} agentName
   * @returns {{ allowed: boolean, tokensRemaining: number, retryAfterMs: number }}
   */
  async tryAcquire(agentName) {
    const agentConfig = await this.getAgentConfig(agentName);
    const key = `ratelimit:bucket:${agentName}`;
    const now = Date.now();

    try {
      const result = await this.redis.eval(
        this.luaScript,
        1,
        key,
        agentConfig.maxTokens,
        agentConfig.refillRate,
        now
      );

      const allowed = result[0] === 1;
      const tokensRemaining = result[1];
      const retryAfterMs = result[2];

      // Track stats
      await this._recordAttempt(agentName, allowed);

      return { allowed, tokensRemaining, retryAfterMs };
    } catch (err) {
      console.error("RateLimiter error:", err.message);
      // Fail-open: allow request if rate limiter errors
      return { allowed: true, tokensRemaining: -1, retryAfterMs: 0 };
    }
  }

  /**
   * Get rate limit config for a specific agent.
   * Falls back to global defaults if no per-agent config exists.
   */
  async getAgentConfig(agentName) {
    try {
      const stored = await this.redis.hget(this.configKey, agentName);
      if (stored) {
        return JSON.parse(stored);
      }
    } catch {
      // Fall through to defaults
    }

    return {
      maxTokens: config.rateLimitMaxTokens,
      refillRate: config.rateLimitRefillRate,
      maxQueueSize: config.rateLimitMaxQueueSize,
      maxWaitTime: config.rateLimitMaxWaitTime,
    };
  }

  /**
   * Set rate limit config for a specific agent.
   */
  async setAgentConfig(agentName, newConfig) {
    const current = await this.getAgentConfig(agentName);
    const merged = { ...current, ...newConfig };

    await this.redis.hset(this.configKey, agentName, JSON.stringify(merged));

    // Reset the bucket so new config takes effect immediately
    await this.redis.del(`ratelimit:bucket:${agentName}`);

    return merged;
  }

  /**
   * Get all agent configs.
   */
  async getAllConfigs() {
    const configs = await this.redis.hgetall(this.configKey);
    const result = {};

    for (const [agent, configStr] of Object.entries(configs)) {
      result[agent] = JSON.parse(configStr);
    }

    return result;
  }

  /**
   * Get the current token bucket status for an agent.
   */
  async getBucketStatus(agentName) {
    const key = `ratelimit:bucket:${agentName}`;
    const bucket = await this.redis.hgetall(key);
    const agentConfig = await this.getAgentConfig(agentName);

    if (!bucket.tokens) {
      return {
        agent: agentName,
        tokens: agentConfig.maxTokens,
        maxTokens: agentConfig.maxTokens,
        refillRate: agentConfig.refillRate,
        status: "full",
      };
    }

    // Calculate current tokens with refill
    const now = Date.now();
    const elapsed = (now - parseFloat(bucket.lastRefill)) / 1000;
    const currentTokens = Math.min(
      agentConfig.maxTokens,
      parseFloat(bucket.tokens) + elapsed * agentConfig.refillRate
    );

    return {
      agent: agentName,
      tokens: Math.round(currentTokens * 100) / 100,
      maxTokens: agentConfig.maxTokens,
      refillRate: agentConfig.refillRate,
      status: currentTokens >= 1 ? "available" : "exhausted",
    };
  }

  /**
   * Get rate limiter stats for an agent.
   */
  async getStats(agentName) {
    const key = `ratelimit:stats:${agentName}`;
    const stats = await this.redis.hgetall(key);

    const allowed = parseInt(stats.allowed || "0", 10);
    const denied = parseInt(stats.denied || "0", 10);
    const total = allowed + denied;

    return {
      agent: agentName,
      allowed,
      denied,
      total,
      denyRate: total > 0 ? ((denied / total) * 100).toFixed(2) : "0.00",
    };
  }

  /**
   * Get stats for all agents.
   */
  async getAllStats() {
    // Scan for all stat keys
    const stats = {};
    let cursor = "0";

    do {
      const [nextCursor, keys] = await this.redis.scan(
        cursor, "MATCH", "ratelimit:stats:*", "COUNT", 100
      );
      cursor = nextCursor;

      for (const key of keys) {
        const agentName = key.replace("ratelimit:stats:", "");
        stats[agentName] = await this.getStats(agentName);
      }
    } while (cursor !== "0");

    return stats;
  }

  /**
   * Reset stats for an agent.
   */
  async resetStats(agentName) {
    await this.redis.del(`ratelimit:stats:${agentName}`);
  }

  // --- Private ---

  async _recordAttempt(agentName, allowed) {
    const key = `ratelimit:stats:${agentName}`;
    const field = allowed ? "allowed" : "denied";
    await this.redis.hincrby(key, field, 1);
  }
}

module.exports = RateLimiter;

File: resilient-layer/rate_limiter.py
import redis.asyncio as redis
import time
import asyncio
import json
from typing import Dict, Any


class TokenBucket:
    """Token bucket algorithm implementation using Redis"""
    
    def __init__(self, redis_client, key_prefix: str, rate: int, window: int):
        self.redis = redis_client
        self.key = f"rate_limit:{key_prefix}"
        self.rate = rate
        self.window = window
    
    async def acquire(self) -> bool:
        """Try to acquire a token. Returns True if allowed, False if rate limited."""
        
        now = time.time()
        window_start = now - self.window
        
        # Lua script for atomic token bucket operation
        lua_script = """
        local key = KEYS[1]
        local now = tonumber(ARGV[1])
        local window_start = tonumber(ARGV[2])
        local max_tokens = tonumber(ARGV[3])
        local window = tonumber(ARGV[4])
        
        redis.call('ZREMRANGEBYSCORE', key, 0, window_start)
        local current_count = redis.call('ZCARD', key)
        
        if current_count < max_tokens then
            local member = now .. '-' .. math.random()
            redis.call('ZADD', key, now, member)
            redis.call('EXPIRE', key, window * 2)
            return 1
        else
            return 0
        end
        """
        
        result = await self.redis.eval(
            lua_script,
            1,
            self.key,
            now,
            window_start,
            self.rate,
            self.window
        )
        
        return result == 1
    
    async def current_usage(self) -> int:
        """Get current token usage"""
        now = time.time()
        window_start = now - self.window
        await self.redis.zremrangebyscore(self.key, 0, window_start)
        return await self.redis.zcard(self.key)
    
    async def reset(self):
        """Reset rate limiter"""
        await self.redis.delete(self.key)


class RateLimiter:
    """Manages multiple token buckets for different agents"""
    
    def __init__(self):
        self.redis = redis.Redis(
            host="redis",
            port=6379,
            decode_responses=True
        )
        self.buckets: Dict[str, TokenBucket] = {}
    
    async def add_agent(self, agent_name: str, rate: int, window: int = 1):
        """Add rate limit configuration for an agent"""
        self.buckets[agent_name] = TokenBucket(
            self.redis,
            f"agent:{agent_name}",
            rate,
            window
        )
    
    async def acquire(self, agent_name: str) -> bool:
        """Check if request is allowed for this agent"""
        if agent_name not in self.buckets:
            return True
        return await self.buckets[agent_name].acquire()
    
    async def get_current_usage(self, agent_name: str) -> int:
        """Get current usage for an agent"""
        if agent_name not in self.buckets:
            return 0
        return await self.buckets[agent_name].current_usage()
    
    async def update_limit(self, agent_name: str, new_rate: int, window: int = 1):
        """Update rate limit dynamically"""
        self.buckets[agent_name] = TokenBucket(
            self.redis,
            f"agent:{agent_name}",
            new_rate,
            window
        )
    
    async def get_stats(self) -> Dict[str, Any]:
        """Get rate limiting statistics"""
        stats = {}
        for agent_name, bucket in self.buckets.items():
            usage = await bucket.current_usage()
            stats[agent_name] = {
                "rate_limit": bucket.rate,
                "window_seconds": bucket.window,
                "current_usage": usage,
                "remaining": max(bucket.rate - usage, 0),
                "utilization_percent": round((usage / bucket.rate) * 100, 2) if bucket.rate > 0 else 0
            }
        return stats
